### DISCUSSION TITLE
AES Encryption with CBC

### DISCUSSION THREAD

**Discussion by anitaa1990** [krzyzanowskim/CryptoSwift#972] [General]
- CryptoSwift version: 0.8.2
- Swift version: 4
- Platform (iOS/macOS/Linux/...): iOS
- Installation method (CocoaPods/Carthage/SPM/...): CocoaPods
- Project name: Finance app

# Description
Hey, I just had a quick question regarding AES encryption. So I get my key from the backend along with the padding. And I need to pass the data I need to encrypt with the key & padding from backend. I use the following code in Android:
```
 public static String encrypt(String key, String algorithm, String value) {
        try {
            SecretKey secretKey = new SecretKeySpec(Base64.decode(key.getBytes(), Base64.NO_WRAP), "AES");
            AlgorithmParameterSpec iv = new IvParameterSpec(Base64.decode(key.getBytes(), Base64.NO_WRAP));
            Cipher cipher = Cipher.getInstance(algorithm);
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, iv);
            return new String(Base64.encode(cipher.doFinal(value.getBytes("UTF-8")), Base64.NO_WRAP));

        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
```

So, let's say:
```
String key = "Q8pTgdOC7h25qKVncJlzDA==";
String padding = "AES/CBC/PKCS5Padding";
```
And when I try to encrypt: 
```
**P@ssword1:**
I get **jYjYDeYBvsIms+JNlBx87g==**
```
This is the corresponding function in JS:
```
let base64Key = CryptoJS.enc.Base64.parse(key);

let encryptedValue = CryptoJS.AES.encrypt(value, base64Key, {
mode: CryptoJS.mode.CBC,
padding: CryptoJS.pad.Pkcs7,
iv: base64Key
});
return encryptedValue.toString();
```

# Steps to reproduce
Now my requirement is that I need to pass the key as iv. But I keep getting 
Thread 1: Fatal error: 'try!' expression unexpectedly raised an error: CryptoSwift.BlockMode.Error.invalidInitializationVector exception.

I understand that the iv needs to be kept within the AES bound size - 16bytes. But not sure how to overcome this. 

Sample code to reproduce the issue:

```swift
import CryptoSwift

// Code goes here
    public class func aesEncrypt(value: String, key: String) -> String {
        let data = value.data(using: .utf8)!
        let encrypted = try! AES(key: key.bytes, blockMode: .CBC(iv: (key.bytes)), padding: .pkcs7).encrypt([UInt8](data))
        let encryptedData = Data(encrypted)
        return encryptedData.base64EncodedString()
    }
```

**Comment by krzyzanowskim** [comment]
You can't overcome the cipher and block mode requirements. IV have to be 16 bytes for AES. If happen your key is 16 bytes as well that would work, otherwise you have to transform it to keep the AES and block mode requirements.

**Comment by krzyzanowskim** [comment]
One option is to use PBKDF to derive key from password, like this:

```swift
let key = try! PKCS5.PBKDF2(password: "secret".bytes, salt: iv.bytes, iterations: 4096, keyLength: 16, variant: .md5).calculate()
```

**Comment by Prabhakaran2702** [comment]
If anyone else is facing the same issue. try with below code.

func aesEncrypt(value: String, key: String) -> String {
let ivSlice = key.bytes[0..<16]
let ivArray = Array<UInt8>(ivSlice)
let data = Array(value.utf8)
    
do {
     let encrypted = try AES(key: ivArray, blockMode: CBC(iv: ivArray), padding: .pkcs5).encrypt(data)

   let encryptedData = Data(encrypted)
  return encryptedData.base64EncodedString()

} catch {
return nil }}
